# Lab 4 - Information Theory in Machine Learning

Welcome to this week's lab on Information Theory! This week, we will dive into the fascinating world of Information Theory as applied to Machine Learning. Specifically, we will focus on two key concepts: Entropy and Information Gain. These principles are fundamental in understanding how decision trees make split decisions to organize data effectively.

### Entropy
- Entropy, in the context of information theory, measures the level of uncertainty or disorder within a set of data.
- In machine learning, particularly in decision trees, entropy helps to determine how a dataset should be split. A high entropy means more disorder, indicating that our dataset is varied. Conversely, low entropy suggests more uniformity in the data.

### Information Gain
- Information Gain measures the reduction in entropy after the dataset is split on an attribute.
- It is crucial in building decision trees as it helps to decide the order of attributes the tree will use for splitting the data. The attribute with the highest Information Gain is chosen as the splitting attribute at each node.

## Part 1: Entropy and Information Gain in Decision Trees
Decision Trees use these concepts to create branches. By choosing splits that maximize Information Gain (or equivalently minimize entropy), a decision tree can effectively categorize data, leading to better classification or regression models.

### Step 1: Import Necessary Libraries

In [6]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree
import matplotlib.pyplot as plt

### Step 2: Load and Explore the Iris Dataset

In [7]:
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target

df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
0         0  
1         0  
2         0  
3         0  
4         0  
..      ...  
145       2  
146       2  
147       2  
148       2  
149       2  

[150 rows x 5 columns]

### Step 3: Calculate Entropy
To calculate the `entropy` we need to:
- First, extract the target variable `y` from your dataset (like the 'target' column in the Iris dataset).
- Then, call `calculate_entropy(y)` to get the entropy.

This function calculates the entropy of a given target variable `y`. It works by first determining the unique classes in `y`, then computes the probability of each class, and uses this probability to calculate the entropy. This is a crucial step in understanding the disorder or uncertainty in the dataset, a fundamental concept in information theory.

In [8]:
def calculate_entropy(y):
    class_labels = np.unique(y)
    entropy = 0
    for label in class_labels:
        probability = len(y[y == label]) / len(y)
        entropy -= probability * np.log2(probability)
    return entropy

calculate_entropy(df['target'])

1.584962500721156

What is your observastion about the calculated Entropy?

The calculated entropy returns *1.584962500721156*. It represents the bit of uncertainty when predicting the species of an Iris flower.

### Step 4: Calculate Information Gain
There are three steps for calculating the Information Gain:
1. Compute Overall Entropy: Use the entropy function from Step 3 on the entire target dataset.
2. Calculate Weighted Entropy for Each Attribute: For each unique value in the attribute, partition the dataset and calculate its entropy. Then calculate the weighted sum of these entropies, where the weights are the proportions of instances in each partition.
3. Compute Information Gain: Subtract the weighted entropy of the split from the original entropy.

The attribute with the highest Information Gain is generally chosen for splitting, as it provides the most significant reduction in uncertainty. This step is critical in constructing an effective decision tree, as it directly influences the structure and depth of the tree.

In [9]:
def calculate_information_gain(df, attribute, target_name):
    total_entropy = calculate_entropy(df[target_name])
    values, counts = np.unique(df[attribute], return_counts=True)
    weighted_entropy = sum((counts[i] / sum(counts)) * calculate_entropy(df.where(df[attribute] == values[i]).dropna()[target_name]) for i in range(len(values)))
    information_gain = total_entropy - weighted_entropy
    return information_gain

sepal_length_ig = calculate_information_gain(df, 'sepal length (cm)', 'target')
print(sepal_length_ig)

sepal_width_ig = calculate_information_gain(df, 'sepal width (cm)', 'target')
print(sepal_width_ig)

petal_length_ig = calculate_information_gain(df, 'petal length (cm)', 'target')
print(petal_length_ig)

petal_width_ig = calculate_information_gain(df, 'petal width (cm)', 'target')
print(petal_width_ig)


0.8769376208910578
0.5166428756804977
1.4463165236458
1.4358978386754417


Discuss your findings here.

From the above results, the attribute with the highest information gain is `petal length (cm)` with 1.4463165236458 information gain. Therefore, this attribute should be the first one to be splitted in order to provide the purest subsets. It is followed by `petal width (cm)` (1.4358978386754417), `sepal length (cm)`(0.8769376208910578) and `sepal width (cm)`(0.5166428756804977).

## Part 2: Apply Entropy and Information Gain on a different dataset

Your task is to choose a new dataset and implement what you learned in `Part 1` on this new dataset.

### Task 1: Implement Entropy and Information Gain

In [4]:
# Selecting the wine dataset from sklearn
from sklearn.datasets import load_wine

wine = load_wine()
df_wine = pd.DataFrame(wine.data, columns=wine.feature_names)

# Set the target variable
df_wine['target'] = wine.target

df_wine.columns

Index(['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium',
       'total_phenols', 'flavanoids', 'nonflavanoid_phenols',
       'proanthocyanins', 'color_intensity', 'hue',
       'od280/od315_of_diluted_wines', 'proline', 'target'],
      dtype='object')

In [14]:
# Calculating the entropy
calculate_entropy(df_wine['target'])

1.5668222768551812

In [ ]:
The entropy returns 1.5668222768551812 in the wine dataset.

In [13]:
# Calculating information gains for each other attributes
alcohol_ig = calculate_information_gain(df_wine, 'alcohol', 'target')
print(f"alcohol: {alcohol_ig}")

malic_acid_ig = calculate_information_gain(df_wine, 'malic_acid', 'target')
print(f"malic_acid: {malic_acid_ig}")

ash_ig = calculate_information_gain(df_wine, 'ash', 'target')
print(f"ash: {ash_ig}")

alcalinity_of_ash_ig = calculate_information_gain(df_wine, 'alcalinity_of_ash', 'target')
print(f"alcalinity_of_ash: {alcalinity_of_ash_ig}")

magnesium_ig = calculate_information_gain(df_wine, 'magnesium', 'target')
print(f"magnesium: {magnesium_ig}")

total_phenols_ig = calculate_information_gain(df_wine, 'total_phenols', 'target')
print(f"total_phenols: {total_phenols_ig}")

flavanoids_ig = calculate_information_gain(df_wine, 'flavanoids', 'target')
print(f"flavanoids, {flavanoids_ig}")

nonflavanoid_phenols_ig = calculate_information_gain(df_wine, 'nonflavanoid_phenols', 'target')
print(f"nonflavanoid_phenols:{ nonflavanoid_phenols_ig}")

proanthocyanins_ig = calculate_information_gain(df_wine, 'proanthocyanins', 'target')
print(f"proanthocyanins:{proanthocyanins_ig}")

color_intensity_ig = calculate_information_gain(df_wine, 'color_intensity', 'target')
print(f"color_intensity:{color_intensity_ig}")

hue_ig = calculate_information_gain(df_wine, 'hue', 'target')
print(f"hue:{hue_ig}")

od280_od315_of_diluted_wines_ig = calculate_information_gain(df_wine, 'od280/od315_of_diluted_wines', 'target')
print(f"od280/od315_of_diluted_wines:{od280_od315_of_diluted_wines_ig}")

proline_ig = calculate_information_gain(df_wine, 'proline', 'target')
print(f"proline:{proline_ig}")

alcohol: 1.3406163077175945
malic_acid: 1.2842014538975586
ash: 0.7863398170147985
alcalinity_of_ash: 0.7357938479915789
magnesium: 0.6361513778259005
total_phenols: 1.1812650269501774
flavanoids, 1.4207548611248442
nonflavanoid_phenols:0.4739066456615457
proanthocyanins:1.1161486280647779
color_intensity:1.3700832318627436
hue:1.0413965047465832
od280/od315_of_diluted_wines:1.3470502797834747
proline:1.3015454175918462


### Task 2: Discuss your findings in detail
Provide detailed explanation and discussion about your findings.

The attribute `flavanoids` (1.4207548611248442) returns the highest information gain in the wine dataset. Which means that it should be the fist node to be splitted to get the purest subset. 

The following nodes with the highest information gain are:
- color_intensity ( 1.3700832318627436), 
- od280/od315_of_diluted_wines (1.3470502797834747), 
- alcohol (1.3406163077175945), 
- proline (1.3015454175918462), 
- total_phenols (1.1812650269501774), 
- malic_acid (1.2842014538975586), 
- proanthocyanins (1.1161486280647779), 
- hue (1.0413965047465832), 
- alcalinity_of_ash (0.7357938479915789), 
- ash (0.7863398170147985), 
- magnesium (0.6361513778259005), 
- nonflavanoid_phenols (0.4739066456615457)

## Submission
Submit a link to your completed Jupyter Notebook file hosted on your private GitHub repository through the submission link in Blackboard.